### ISSUE TITLE
Cannot access mjd property on Time with out_subfmt set to date 

### ISSUE THREAD

**Issue by StuartLittlefair** [astropy/astropy#9810]
### Description

Accessing the ```mid``` property on a ```Time``` with custom ```out_subfmt``` set raises an exception.

### Steps to Reproduce

```python
In [3]: from astropy.time import Time                                                                                

In [4]: t = Time('2019-12-20', out_subfmt='date_hm')                                                                 

In [5]: t.mjd                                                                                                        
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-5-bd7732b99f6e> in <module>
----> 1 t.mjd

~/anaconda3/envs/astropy_est/lib/python3.7/site-packages/astropy/time/core.py in __getattr__(self, attr)
   1694 
   1695         elif attr in self.FORMATS:
-> 1696             return self.to_value(attr, subfmt=None)
   1697 
   1698         elif attr in TIME_SCALES:  # allowed ones done above (self.SCALES)

~/anaconda3/envs/astropy_est/lib/python3.7/site-packages/astropy/time/core.py in to_value(self, format, subfmt)
    999                     raise ValueError('subformat must match one of {}'
   1000                                      .format([option[0] for option in
-> 1001                                               tm._time.subfmts])) from None
   1002                 else:
   1003                     raise

ValueError: subformat must match one of ['float', 'long', 'decimal', 'str', 'bytes']
```

### System Details

Darwin-19.2.0-x86_64-i386-64bit
Python 3.7.4 (default, Aug 13 2019, 15:17:50) 
[Clang 4.0.1 (tags/RELEASE_401/final)]
Numpy 1.17.4
Scipy 1.3.1
astropy 4.1.dev27242

**Comment by StuartLittlefair** [comment]
Not sure as to the best way to proceed with this. 

The underlying problem is that [accessing t.mjd](https://github.com/astropy/astropy/blob/master/astropy/time/core.py#L1696) uses ```__getattr__``` which calls  ```to_value``` with ```subfmt``` equal to ```None```.

This uses the object's own ```out_subfmt```, which is not compatible with a float. One option is to change [line 1696](https://github.com/astropy/astropy/blob/master/astropy/time/core.py#L1696) to use ```subfmt='*'``` - this uses the first available sub format compatible with the requested attribute. However, this may change the output sub format when the original was compatible with the desired attribute, but wasn't first in the list...

**Comment by pllim** [comment]
cc @taldcroft and @mhvk

**Comment by mhvk** [comment]
@StuartLittlefair - hmm, that's tricky. Obviously, the problem is that `out_subfmt` makes sense for `iso`, for which you created the instance, but not for `format='mjd'`, which gets created to make the `mjd`. I guess when we temporarily create the `TimeMJD` instance, we should remove the subformat. Or we should ignore a subformat if it doesn't exist for the given type (at least for this `__getattr__` route). I'll look into it...

**Comment by mhvk** [comment]
p.s. Unsurprisingly, the following errors too:
```
Time(50000., format='mjd', out_subfmt='str').isot
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-11-7552ff8859d0> in <module>()
----> 1 Time(50000., format='mjd', out_subfmt='str').isot

/data/mhvk/venv/astropy-dev/lib/python3.7/site-packages/astropy-4.1.dev27239-py3.7-linux-x86_64.egg/astropy/time/core.py in __getattr__(self, attr)
   1694 
   1695         elif attr in self.FORMATS:
-> 1696             return self.to_value(attr, subfmt=None)
   1697 
   1698         elif attr in TIME_SCALES:  # allowed ones done above (self.SCALES)

/data/mhvk/venv/astropy-dev/lib/python3.7/site-packages/astropy-4.1.dev27239-py3.7-linux-x86_64.egg/astropy/time/core.py in to_value(self, format, subfmt)
    999                     raise ValueError('subformat must match one of {}'
   1000                                      .format([option[0] for option in
-> 1001                                               tm._time.subfmts])) from None
   1002                 else:
   1003                     raise

ValueError: subformat must match one of ['date_hms', 'date_hm', 'date']
```

**Comment by taldcroft** [comment]
This is fixed with the 2-line patch below.  Probably need to check for other places like this.
```
diff --git a/astropy/time/core.py b/astropy/time/core.py
index 2d7bbf92a..b594b3c14 100644
--- a/astropy/time/core.py
+++ b/astropy/time/core.py
@@ -983,6 +983,8 @@ class Time(ShapedLikeNDArray):
                 tm = self
             else:
                 tm = self.replicate(format=format)
+                if hasattr(tm._time, 'subfmts') and tm.out_subfmt not in tm._time.subfmts:
+                    tm.out_subfmt = '*'
 
             # Custom TimeFormat subclasses created in astropy versions <= 4.0
             # may not be able to handle being passes on a out_subfmt.
```

But there is one test failure that is related, so this isn't the whole story but probably close.

**Comment by mhvk** [comment]
Yes I was thinking along those lines as well. I think you'd need to do the checking a bit more carefully since the `subfmt` can include a `*`. Also, perhaps this is better done inside `TimeFormat`? (its initializer could be told to be allowed to throw away `subfmt` that didn't exist)

**Comment by taldcroft** [comment]
Fixed by #9812, hopefully.
